mlflow/utils/rest_utils.py

```diff
@@ -31,6 +31,31 @@
         504,  # Gateway Timeout
     ]
 )
+# A dictionary caching Requests.Sessions during runtime. This is useful for TCP connection re-use.
+_REQUEST_SESSIONS = {}
+
+
+def _get_request_session(retry_kwargs):
```

Comment:
Can we use [functools.lru_cache](https://docs.python.org/3/library/functools.html#functools.lru_cache)?

Reply:
```python
>>> hash({"a": 0})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'dict'
```

`dict` is

Reply:
So we need a trick like this if we use `lru_cache`:

```python
from functools import lru_cache


class HashableDict(dict):
    def __hash__(self):
        return hash(frozenset(self.items()))


@lru_cache(maxsize=100)
def func(kwargs):
    print(kwargs)


func(HashableDict({"a": 0}))
func(HashableDict({"a": 0}))
func(HashableDict({"a": 0, "b": 1}))
```

Reply:
Won't we have to set a cache expiry time in order to fetch updated page results server-side and prevent stale cached data from being displayed if a previous request's data updates? (I imagine this might be applicable to certain pages and not others, though)

`@lru_cache(maxsize=<x>, seconds=30)` (or some other reasonable time period based on updating periodicities of run data while navigating). 

Would we want to cache Put requests or just Get requests?

Potential future food for thought for this: [requests-cache](https://requests-cache.readthedocs.io/en/latest/user_guide.html) which can give endpoint isolation for caching certain endpoints with different settings and a bunch of other potentially useful features to lessen the tracking server burden.

Reply:
Not sure if I fully understand your concern, but the added function LRU only caches the `requests.Session` object. All outgoing requests (initiated by `session.request`) are not cached and should never be.

The objective here is to persist the TCP channel across HTTP requests to the same destination IP/Port combo, nothing else.

Reply:
I was just curious if a NAT router / Firewall change between the server and the client introducing a network change would lead a long-running session into defaulting to maximum backoff retry event waiting before invalidating the cache value and establishing a new TCP connection. Maybe I'm thinking about this wrong though.

Reply:
See the top section of this page: https://docs.python-requests.org/en/latest/user/advanced/#session-objects
I think we can safely leave the network management and TCP connection re-use to urllib3, which is managed by the session object.

Reply:
100%. I forgot about how requests will drop the connection and re-validate with a new connection from the pool if a connection cannot be established.